File: octave/Table1.m
% Table 1 in Section 8.1 in Canay, Illanes and Velez (2023)

% Inputs

% Amatrix200701_fake.mat                 fake data
%  - A_matrix            n  x (1+J0)     matrix of revenue differential
%  - D_matrix            n  x (1+J)      matrix of all product portfolio
%  - J0_vec              J0 x 2          matrix of product ownership by two firms

% G_restriction.m                        find test statistic and c. value
%  - m_function                          compute (26)-(27)
%  - m_hat                               compute a version of (38)
%  - cvalue_SN2S                         compute c. value as in (41)
%  - cvalue_EB2S                         compute c. value as in (48)

% output

% 'table1.tex'                           confidence intervals and comp.time

% comment:
% the first column of A_matrix and D_matrix were used to index the markets,
% these first columns are useless in the rest of the code.

clc; clear all; close all;

addpath('1_functions');
load('Amatrix200701_fake.mat');
mkdir('_results');

%% 1 Setup

dgp = struct;
dgp.A_matrix = A_matrix;
dgp.D_matrix = D_matrix;

dgp.J0_vec = J0_vec; % (product, firm), where firm = 1 coca-cola, firm=2 energy-products
dgp.num_market = size(A_matrix, 1);
dgp.num_product = size(D_matrix, 1) - 1;

settings = struct;
settings.Vbar = {500, 1000}; % Vbar is defined in Assumption 4.2 and appears in eq. (26)-(27).
settings.IV = {'N'}; % No instrumental variables
settings.test_stat = {'CCK'}; % CCK as in eq. (38).
settings.cv = {'SN2S', 'EB2S'}; % Critical values as in eq. (41) and (47).
settings.alpha = 0.05; % significance level
settings.grid = {1, 2}; % compute moment functions for 1: theta1, 2: theta2

sim = struct;
sim.grid_Theta1 = linspace(-40, 100, 1401)';
sim.grid_Theta2 = linspace(-40, 100, 1401)';
sim.rng_seed = 20220826;
sim.num_boots = 1000;
sim.rng_seed_R = 20220818;
sim.num_boots_R = 250;
sim.sim_name = 'Table1_0423fake';

specs = cell(4, 1);

specs{1} = {settings.Vbar{1}, settings.IV{1}, ...
                                              settings.test_stat{1}, settings.cv{1}}; % Vbar=500, IV=N, test0=CCK, cvalue = SN2S,

specs{2} = {settings.Vbar{1}, settings.IV{1}, ...
                                              settings.test_stat{1}, settings.cv{2}}; % Vbar=500, IV=N, test0=CCK, cvalue = EB2S,

specs{3} = {settings.Vbar{2}, settings.IV{1}, ...
                                              settings.test_stat{1}, settings.cv{1}}; % Vbar=1000, IV=N, test0=CCK, cvalue= SN2S,

specs{4} = {settings.Vbar{2}, settings.IV{1}, ...
                                              settings.test_stat{1}, settings.cv{2}}; % Vbar=1000, IV=N, test0=CCK, cvalue= EB2S,

results = struct;
results.CI1_vec = zeros(4, 2); % specs x {LB, UB}
results.CI2_vec = zeros(4, 2);
results.comp_time = zeros(4, 1);

results.Tn_vec1 = zeros(size(sim.grid_Theta1, 1), 4);
results.Tn_vec2 = zeros(size(sim.grid_Theta2, 1), 4);

%% 2 Computation
%  two steps:
%             i) compute test statistic and critical value
%            ii) construct confidence interevals

W_data = D_matrix(:, 2:end);
alpha_input = settings.alpha;
num_boots = sim.num_boots;

for sim0 = 1:4

    disp(['case ' num2str(sim0)])

    Vbar = specs{sim0}{1};
    IV = specs{sim0}{2};
    test0 = specs{sim0}{3};
    cvalue = specs{sim0}{4};

    reject_H1 = zeros(size(sim.grid_Theta1, 1), 1); % save which points reject H0
    reject_H2 = zeros(size(sim.grid_Theta2, 1), 1); % save which points reject H0

    Test1_vec = zeros(size(sim.grid_Theta1, 1), 1);
    cv1_vec = zeros(size(sim.grid_Theta1, 1), 1);
    Test2_vec = zeros(size(sim.grid_Theta2, 1), 1);
    cv2_vec = zeros(size(sim.grid_Theta2, 1), 1);

    % Step 1: find test stat. Tn(theta) and c.value(theta) using G_restriction

    tic

    for point0 = 1:size(sim.grid_Theta1, 1)
        theta0 = [sim.grid_Theta1(point0, :) 0]';
        grid0 = settings.grid{1};

        %test_H0: [T_n, c_value]
        test_H0 = G_restriction(W_data, A_matrix, theta0, J0_vec, Vbar, IV, grid0, test0, cvalue, alpha_input, num_boots);

        Test1_vec(point0) = test_H0(1);
        cv1_vec(point0) = test_H0(2);

        reject_H1(point0) = 1 * (test_H0(1) > test_H0(2));
    end

    for point0 = 1:size(sim.grid_Theta2, 1)
        theta0 = [0 sim.grid_Theta2(point0, :)]';
        grid0 = settings.grid{2};

        %test_H0: [T_n, c_value]
        test_H0 = G_restriction(W_data, A_matrix, theta0, J0_vec, Vbar, IV, grid0, test0, cvalue, alpha_input, num_boots);

        Test2_vec(point0) = test_H0(1);
        cv2_vec(point0) = test_H0(2);

        reject_H2(point0) = 1 * (test_H0(1) > test_H0(2));
    end

    results.Tn_vec1(:, sim0) = Test1_vec;
    results.Tn_vec2(:, sim0) = Test2_vec;

    % Step 2: find confidence intervals using Tn(theta) and c.value(theta)

    % Confidence Interval for theta1

    CS_vec = [];

    for point0 = 1:size(sim.grid_Theta1, 1)
        theta1 = sim.grid_Theta1(point0, :)';

        if reject_H1(point0) == 0
            CS_vec = [CS_vec; theta1'];
        end

    end

    if sum(size(CS_vec)) == 0 % it may be the CI is empty
        results.CI1_vec(sim0, :) = [NaN NaN];
        [~, point0] = min(Test1_vec);
        theta1 = sim.grid_Theta1(point0, :);
        results.CI1_vec(sim0, 2) = theta1; % in this case, we report [nan, argmin test statistic]
    else
        results.CI1_vec(sim0, :) = [min(CS_vec) max(CS_vec)];
    end

    % Confidence Interval for theta2

    CS_vec = [];

    for point0 = 1:size(sim.grid_Theta2, 1)
        theta2 = sim.grid_Theta2(point0, :)';

        if reject_H2(point0) == 0
            CS_vec = [CS_vec; theta2'];
        end

    end

    if sum(size(CS_vec)) == 0 % it may be the CI is empty
        results.CI2_vec(sim0, :) = [NaN NaN];
        [~, point0] = min(Test2_vec);
        theta2 = sim.grid_Theta1(point0, :);
        results.CI1_vec(sim0, 2) = theta2; % in this case, we report [nan, argmin test statistic]
    else
        results.CI2_vec(sim0, :) = [min(CS_vec) max(CS_vec)];
    end

    toc
    time = toc;
    results.comp_time(sim0, 1) = time;
end

%% 3 Save results
save(fullfile('_results', strcat(sim.sim_name, '.mat')), 'dgp', 'settings', 'sim', 'results', 'specs');

%% 4 Print table
cd_name = 'tables-tex';
mkdir(fullfile('_results', cd_name));

f = fopen(fullfile('_results', cd_name, strcat(sim.sim_name, 'table1.tex')), 'w'); % Open file for writing

fprintf(f, '%s\n', '\begin{tabular}{c c c c c}')
fprintf(f, '%s\n', '\hline \hline')
fprintf(f, '%s\n', '~ & Crit. Value & $\theta_1$: Coca-Cola &$\theta_2$: Energy Brands & Comp. Time \\')
fprintf(f, '%s\n', '\hline')

for row0 = 1:size(specs, 1)

    Vbar = specs{row0}{1};
    cvalue = specs{row0}{4};

    if strcmp(cvalue, 'SN2S')
        Vbar = ['$\Bar{V}$=' num2str(Vbar)];
        cvalue0 = 'self-norm';
    elseif strcmp(cvalue, 'EB2S')
        Vbar = "~";
        cvalue0 = 'bootstrap';
    end

    fprintf(f, '%s%s%s%s%5.1f%s%5.1f%s%5.1f%s%5.1f%s%5.1f%s\n', ...
        Vbar, ' & ', cvalue0, ' & [', ...
        results.CI1_vec(row0, 1), ' , ', results.CI1_vec(row0, 2), '] & [', ...
        results.CI2_vec(row0, 1), ' , ', results.CI2_vec(row0, 2), '] &', ...
        results.comp_time(row0, 1), '\\');

    if row0 == 2
        fprintf(f, '%s\n', '\hline')
    end

end

fprintf(f, '%s\n', '\hline \hline')
fprintf(f, '%s', '\end{tabular}')

File: python/table_1.py
""" Table 1 in Section 8.1 in Canay, Illanes and Velez (2023)
"""

import time
from pathlib import Path

import numpy as np
import texttable as tt
from joblib import Parallel, delayed
from latextable import draw_latex

import ineq_functions as ineq

# Make folders for results
results_dir = Path(__file__).resolve().parent / "_results"
tables_dir = results_dir / "tables-tex"
tables_dir.mkdir(parents=True, exist_ok=True)


def load_data(name):
    file_path = Path(__file__).resolve().parent.parent / "data" / (name + ".csv")
    return np.loadtxt(file_path, delimiter=",")


dgp = {i: load_data(i) for i in ["A", "D", "J0"]}
dgp["W"] = dgp["D"][:, 1:]
n = dgp["A"].shape[0]

settings = {
    "v_bar": [500, 500, 1000, 1000],
    "test_stat": ["CCK", "CCK", "CCK", "CCK"],
    "cv": ["SN2S", "EB2S", "SN2S", "EB2S"],
    "alpha": 0.05,
    "iv": None,
}

sim = {
    "grid_size": 1401,
    "rng_seed": 20220826,
    "bootstrap_replications": 1000,
    "num_robots": 4,
    "sim_name": "table_1",
}
sim["grid_theta"] = (
    np.linspace(-40, 100, sim["grid_size"]),
    np.linspace(-40, 100, sim["grid_size"]),
)

results = {
    "ci_vector": (np.empty((4, 2)), np.empty((4, 2))),
    "tn_vector": (np.empty((sim["grid_size"], 4)), np.empty((sim["grid_size"], 4))),
    "comp_time": np.empty(4),
}

# Generate bootstrap indices
bootstrap_indices = ineq.helpers.get_bootstrap_indices(
    n, sim["bootstrap_replications"], sim["rng_seed"]
)

for sim_i in range(4):
    print("Simulation:", sim_i + 1)
    # Obtain the time at the beginning of the simulation
    tic = time.perf_counter()
    for theta_index in range(2):
        # Step 1: find test stat. Tn(theta) and c.value(theta) using G_restriction
        def theta0(theta, theta_index):
            the_theta = np.zeros(2)
            the_theta[theta_index] = theta
            return the_theta

        output = np.array(
            Parallel(n_jobs=sim["num_robots"])(
                delayed(ineq.g_restriction)(
                    theta=theta0(theta, theta_index),
                    w_data=dgp["W"],
                    a_matrix=dgp["A"],
                    j0_vector=dgp["J0"],
                    v_bar=settings["v_bar"][sim_i],
                    alpha=settings["alpha"],
                    grid0=theta_index + 1,
                    iv_matrix=settings["iv"],
                    test0=settings["test_stat"][sim_i],
                    cvalue=settings["cv"][sim_i],
                    bootstrap_indices=bootstrap_indices,
                )
                for theta in sim["grid_theta"][theta_index]
            )
        )

        # Because g_function returns a list, and Parallel returns it's runs in
        # a list, test_vec and cv_vec are the columns of the output matrix!
        test_vec = output[:, 0]
        cv_vec = output[:, 1]

        # Theta values for which the null is not rejected
        cs_vec = sim["grid_theta"][theta_index][test_vec <= cv_vec]

        # Create results objects
        results["tn_vector"][theta_index][:, sim_i] = test_vec

        if len(cs_vec) == 0:
            # it may be the CI is empty
            results["ci_vector"][theta_index][sim_i,] = [
                np.NaN,
                sim["grid_theta"][theta_index][test_vec.argmin()],
            ]
        else:
            results["ci_vector"][theta_index][sim_i,] = [cs_vec.min(), cs_vec.max()]

    # Stop the timer
    toc = time.perf_counter()
    results["comp_time"][sim_i] = toc - tic
    print("~> time:", results["comp_time"][sim_i])

# Save results
(results_dir / sim["sim_name"]).mkdir(exist_ok=True)
for key, value in results.items():
    np.save(results_dir / sim["sim_name"] / key, value)

# Make and print the table
tableObj = tt.Texttable(0)

tableObj.set_cols_align(["l", "l", "c", "c", "c"])
tableObj.set_cols_dtype(["i", "t", "t", "t", "f"])

the_table = np.array(settings["v_bar"])
the_table = np.column_stack((the_table, settings["cv"]))
for ci_theta in results["ci_vector"]:
    the_table = np.column_stack(
        (
            the_table,
            [
                "[" + "{:.1f}".format(x[0]) + ", " + "{:.1f}".format(x[1]) + "]"
                for x in ci_theta
            ],
        )
    )
the_table = np.column_stack((the_table, results["comp_time"]))
the_table = np.vstack(
    (
        [
            "$\\Bar{V}$",
            "Crit. Value",
            "$\\theta_1$: Coca-Cola",
            "$\\theta_2$: Energy Brands",
            "Comp. Time",
        ],
        the_table,
    )
)

tableObj.add_rows(the_table)
print(tableObj.draw())

# Save table
with open(tables_dir / (sim["sim_name"] + ".tex"), "w", encoding="utf8") as f:
    f.write(draw_latex(tableObj))
